chainer/functions/activation/relu.py

```diff
@@ -17,40 +18,70 @@ class ReLU(function_node.FunctionNode):
     """Rectified Linear Unit."""
 
     _use_cudnn = False
+    _ideep_hint = None
 
     def check_type_forward(self, in_types):
         type_check.expect(
             in_types.size() == 1,
             in_types[0].dtype.kind == 'f',
         )
 
-    def forward_cpu(self, x):
+    def forward_cpu(self, inputs):
+        x, = inputs
+        if (x.dtype == numpy.float32
+                and (x.ndim == 2 or x.ndim == 4)
+                and _ideep.should_use_ideep('>=auto')):
+
+            # iDeep implementation
+            return self.forward_ideep(inputs)
+
         self.retain_outputs((0,))
-        return utils.force_array(numpy.maximum(x[0], 0, dtype=x[0].dtype)),
+        return utils.force_array(numpy.maximum(x, 0, dtype=x.dtype)),
+
+    def forward_ideep(self, inputs):
+        self.retain_inputs((0,))
+        self.retain_outputs((0,))
+
+        cc = _ideep.ideep.xnn.ReLUForward(inputs)
+        self._ideep_hint = cc.hint
 
-    def forward_gpu(self, x):
-        if chainer.should_use_cudnn('==always') and x[0].flags.c_contiguous:
+        y, = cc.execute_on()
+        return y,
+
+    def forward_gpu(self, inputs):
+        x, = inputs
+        if chainer.should_use_cudnn('==always') and x.flags.c_contiguous:
             # cupy.activation_backward requires the input.
             # So, we retain it for backward computation.
             self.retain_inputs((0,))
             self._use_cudnn = True
-            y = cudnn.activation_forward(x[0], _mode)
+            y = cudnn.activation_forward(x, _mode)
         else:
-            y = cuda.cupy.maximum(x[0], 0)
+            y = cuda.cupy.maximum(x, 0)
         self.retain_outputs((0,))
         return y,
 
-    def backward(self, indexes, gy):
-        y = self.get_retained_outputs()[0]
+    def backward(self, indexes, grad_outputs):
+        gy, = grad_outputs
+        y, = self.get_retained_outputs()
+        if self._ideep_hint is not None:
+            # iDeep implementation
+            x, = self.get_retained_inputs()
+            return ReLUGradIdeep(x, y, self._ideep_hint).apply((gy,))
         if chainer.should_use_cudnn('==always') and self._use_cudnn:
-            x = self.get_retained_inputs()[0]
-            return ReLUGrad3(x, y).apply((gy[0],))
-        else:
-            return ReLUGrad2(y).apply((gy[0],))
+            # cuDNN implementatioin
+            x, = self.get_retained_inputs()
+            return ReLUGradCudnn(x, y).apply((gy,))
+        # Generic implementaion
+        return ReLUGrad2(y).apply((gy,))
 
 
 def _heaviside(x):
-    return (x > 0).astype(x.dtype)
+    if _ideep.is_available() and isinstance(x, _ideep.ideep.mdarray):
+        # ideep.mdarray does not support __gt__ yet
+        return numpy.greater(x, 0).astype(x.dtype)
```

Comment:
iDeep document says `'MD-Array' supports python new buffer protocol and operates compatibily with NumPY ND-Array.`, so maybe we can ask iDeep team to support operator overloads?

Currently mdarray is using `tp_getattro` to fallback attribute access to numpy, however, special method lookup (e.g., `__add__` for `+`) does not go through this route. https://docs.python.org/3/c-api/typeobj.html#c.PyTypeObject.tp_getattro
Maybe proxy method or in python layer is needed (by them, or by us?)